python/dgl/dataloading/pytorch/dataloader.py

```diff
@@ -101,6 +101,32 @@ def __init__(self, dataset, shuffle=False, batch_size=1,
                 self.num_samples = math.ceil(len(self.dataset) / self.num_replicas)  # type: ignore
             self.total_size = self.num_samples * self.num_replicas
 
+            if self.rank == 0:
+                # pad the indices
+                indices = th.arange(len(self.dataset))
+                if not self.drop_last:
+                    # add extra samples to make it evenly divisible
+                    indices = th.cat([indices, indices[:(self.total_size - indices.shape[0])]])
+                else:
+                    # remove tail of data to make it evenly divisible.
+                    indices = indices[:self.total_size]
+                assert indices.shape[0] == self.total_size
+
+                # share indices among processes
+                indices = indices.share_memory_()
+                _shared_filename = list(indices.storage()._share_filename_())
+            else:
+                _shared_filename = [None, None, None]
+            # Note: dist.broadcast_object_list() is a new feature introduced in PyTorch 1.7.0.
+            #       We may need a fallback for lower PyTorch versions.
+            dist.broadcast_object_list(_shared_filename, src=0)
+            # reconstruct the shared tensor
+            if self.rank != 0:
+                indices = th.LongTensor(th.LongStorage._new_shared_filename(*_shared_filename))
```

Comment:
Interesting usage that I don't know before. Can we also use this to share the graph across multiple process? @BarclayII

Reply:
Since they are prefixed with `_` I suspect that they are undocumented hence unstable.  Could we use a queue instead since it's the preferred way of sharing tensors across processes?

Reply:
An alternative might be using `NDArray::EmptyShared` which creates a shared memory tensor directly.  The shared memory name can be generated in rank 0 and broadcasted to other ranks.

Reply:
>  use a queue

We have to create a queue and pass it to the worker processes when spawning/forking them, hard to do it inside the dataloader.

> using `NDArray::EmptyShared`

In fact I have tried it. `empty_shared_mem` uses a name to identify the shared memory. But I realize that there may be multiple dataloaders for training/test/validation, so we cannot use a hard code here.

@BarclayII

Reply:
> `empty_shared_mem` uses a name to identify the shared memory. But I realize that there may be multiple dataloaders for training/test/validation, so we cannot use a hard code here.

That's because you use the same name `_shared_indices_for_ddp`?  I think you could generate a name in rank 0 and distribute it with `dist.broadcast` like you are doing here.

Reply:
I think it's fine to use this internal API since the semantic is clear. We can add some check or fallback solution if needed.
Our EmptyShared doesn't support windows as what I remembered. That will break the user experience on windows.

Reply:
Also I believe for error handling and error messages, Pytorch does a better job than us

Reply:
You could also partition the training indices without shuffling to every rank, and then shuffle each partition individually.

Reply:
@VoVAllen I can take care of supporting shared memory on Windows since they have a very similar interface.

Reply:
> You could also partition the training indices without shuffling to every rank, and then shuffle each partition individually.

Do you mean using a static split, i.e., the partition doesn't change between epochs? For this case, I think letting user partition the indices and initialize the dataloader with `use_ddp=False` will make things much easier.

If you mean something like:
```python
self._indices = _shared_indices[self.rank:self.total_size:self.num_replicas]
# then in iter_ddp():
np.random.shuffle(self._indices.numpy())
```
This would be even slower, which takes 2x times for shuffling compared to the above `randperm and assign`.

Reply:
> Do you mean using a static split, i.e., the partition doesn't change between epochs? For this case, I think letting user partition the indices and initialize the dataloader with `use_ddp=False` will make things much easier.

Right, say that we have N nodes and R workers, then each rank `r` is only responsible for shuffling the node ID `[N // R * r, N // R * (r+1)]`.  You could still use `randperm` and assign.  This should be OK because even if the nodes are ordered by label, since the gradients are synchronized across workers and would not be too far off.

Reply:
@BarclayII I have pushed a version that uses a static indices split for each process. It's less random but being the fastest way.